### ISSUE TITLE
Inconsistent equality behavior between Series and Index

### ISSUE THREAD

**Issue by ForeverWintr** [InvestmentSystems/static-frame#126]
SF will align series indexes before checking equality. This works as expected:

```python
sf.Series([1, 2]) == sf.Series([1, 2, 3])
<Series>
<Index>
0        True
1        True
2        False
<int64>  <bool>
```

However, sf.Indexes behave differently:

```python
sf.Series([1, 2]).index == sf.Series([1, 2, 3]).index
array([False, False])
```
Note that the resulting boolean array only has length 2, and I would expect `array([True, True, False])`

Possibly related to the above, trying to compare a series to an index only works if they have the same length:

```python
sf.Series([0, 1, 3]) == sf.Series([1, 2, 3]).index
<Series>
<Index>
0        True
1        True
2        False
<int64>  <bool>
```

```python
sf.Series([0, 1]) == sf.Series([1, 2, 3]).index
# static_frame.core.exception.ErrorInitSeries: values and index do not match length
```

I'm not sure what the desired behavior is here, but from my naive perspective it would make most sense for index comparison to behave the same as series comparison. Are there other considerations that make this difficult/unwise?

**Comment by flexatone** [comment]
Thanks for posting this issue!

In general, Index operations with operators fall back on NumPy operations of the array of labels stored in the `Index`. So for example, if we have a numeric Index  (and such operations only work with a numeric `Index`), we get the following:

```python
>>> sf.Index([1, 2, 3]) * 2
array([2, 4, 6]
```
The case you isolate is not what we get from NP arrays alone:

```python
>>> np.array([1, 2]) == np.array([1,2,3])
False
```
In order for `Index`-to-`Index` operations to work as `Series`, or `Index` to `Series` comparisons, we would need to reindex the `Index` to the union index of both operands. I suppose this is possible, but is awkward as it could introduce NaNs in the labels of the index.

But what would be returned be these operations: do you think a `Index` == `Index` operation should return a `Series`? That is what would be necessary to capture the Boolean result. That seems odd, but is possible.

Maybe operator functionality should not be exposed for `Index` objects at all.

**Comment by ForeverWintr** [comment]
It makes sense that we don't want to introduce nans in an index (side note: should SF disallow nan in indexes?), but it's probably acceptable to return a different data type containing nans.

Two options I can think of: 

- return a Series, whose index is the union index.
- return a numpy array of bools the size of the union index. 

The latter is basically what I expected to happen. 

Not exposing equality between indexes is probably ok too. In that case users would have to do:

```python
sf.Series(sf.Series([1, 2]).index) == sf.Series(sf.Series([1, 2, 3]).index)
<Series>
<Index>
0        True
1        True
2        False
<int64>  <bool>
```

That may be more explicit, especially if that's how SF would implement the operation under the hood.

**Comment by flexatone** [comment]
The general approach to handling this, which will be implemented on 0.7, is that binary operators on `Index` will operate as 1D array do in NumPy, with the exception that different sized comparison with `==`  will raise a `ValueError`, as NumPy does for other binary operators (but not `==` and `!=`).

This inconsistency in NumPy is shown below:
```
>>> np.array([3, 4]) == np.array([1, 2, 3])
False
>>> np.array([3, 4]) + np.array([1, 2, 3])
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (2,) (3,) 
```

**Comment by ForeverWintr** [comment]
That sounds like a good solution. Thanks!

**Comment by flexatone** [comment]
This is now implemented for 0.7. Fixing this also exposed places where other containers were falling back on this behavior, i.e., when NumPy returns a Boolean when applying a binary operator. In all cases these now raise.

```python
>>> sf.Series([1, 2]).index == sf.Series([1, 2, 3]).index
ValueError: operands could not be broadcast together

>>> sf.Index([3, 4]) == sf.Index([3, 4])
array([ True,  True])

>>> sf.Index([3, 4]) == sf.Index([3, 4, 5])
ValueError: operands could not be broadcast together
```
